src/detect-asn1.c

```diff
@@ -148,15 +148,37 @@ static int DetectAsn1Match(DetectEngineThreadCtx *det_ctx, Packet *p,
     }
 
     const DetectAsn1Data *ad = (const DetectAsn1Data *)ctx;
-    int32_t offset;
+    uint16_t offset;
     if (ad->flags & ASN1_ABSOLUTE_OFFSET) {
         offset = ad->absolute_offset;
     } else if (ad->flags & ASN1_RELATIVE_OFFSET) {
-        offset = ad->relative_offset;
+        // relative offset in regards to the last content match
+
+        // Range check the relative_offset before downcasting to uint16_t
+        if (ad->relative_offset > UINT16_MAX
+            || ad->relative_offset < -UINT16_MAX) {
+                return 0;
+        }
+
+        // Check for over/underflows
```

Comment:
I find these checks hard to follow. I think I would feel better seeing the checks done as u32 (or i32) and then only down casting after the values have been validated. Easier to follow for me anyway.

Reply:
Agreed, re-reading this I get confused. I'll see if I can come up with a better solution